src/FffGcodeWriter.cpp

```diff
@@ -677,52 +617,116 @@ void FffGcodeWriter::processDraftShield(SliceDataStorage& storage, GCodePlanner&
+        {
+            for (unsigned int used_idx = 0; used_idx < extruder_is_used.size(); used_idx++)
+            {
+                extruder_is_used[used_idx] = true;
+            }
+        }
+    }
+    for (unsigned int extruder_nr = 0; extruder_nr < extruder_count; extruder_nr++)
+    {
+        if (extruder_nr == start_extruder)
         { // skip the current extruder, it's the one we started out planning
             continue;
         }
+        if (!extruder_is_used[extruder_nr])
+        {
+            continue;
+        }
         ret.push_back(extruder_nr);
     }
-    assert(ret.size() == (size_t)extruder_count && "All extruders must be planned, even if later it appears one wasn't used.");
+    assert(ret.size() <= (size_t)extruder_count && "Not more extruders may be planned in a layer than there are extruders!");
     return ret;
 }
 
-std::vector<unsigned int> FffGcodeWriter::calculateMeshOrder(SliceDataStorage& storage, int extruder_nr)
+std::vector<unsigned int> FffGcodeWriter::calculateMeshOrder(const SliceDataStorage& storage, int extruder_nr) const
 {
     OrderOptimizer<unsigned int> mesh_idx_order_optimizer;
 
     for (unsigned int mesh_idx = 0; mesh_idx < storage.meshes.size(); mesh_idx++)
     {
-        SliceMeshStorage& mesh = storage.meshes[mesh_idx];
+        const SliceMeshStorage& mesh = storage.meshes[mesh_idx];
         if (mesh.getSettingAsIndex("extruder_nr") == extruder_nr)
         {
-            Mesh& mesh_data = storage.meshgroup->meshes[mesh_idx];
-            Point3 middle = (mesh_data.getAABB().min + mesh_data.getAABB().max) / 2;
+            const Mesh& mesh_data = storage.meshgroup->meshes[mesh_idx];
+            const Point3 middle = (mesh_data.getAABB().min + mesh_data.getAABB().max) / 2;
             mesh_idx_order_optimizer.addItem(Point(middle.x, middle.y), mesh_idx);
         }
     }
     std::list<unsigned int> mesh_indices_order = mesh_idx_order_optimizer.optimize();
+    std::list<unsigned int>::iterator starting_mesh_it = mesh_indices_order.end();
+    { // calculate starting_mesh_it
+        const ExtruderTrain* train = storage.meshgroup->getExtruderTrain(extruder_nr);
+        const Point layer_start_position = Point(train->getSettingInMicrons("layer_start_x"), train->getSettingInMicrons("layer_start_y"));
+        coord_t best_dist2 = std::numeric_limits<coord_t>::max();
+        for (std::list<unsigned int>::iterator mesh_it = mesh_indices_order.begin(); mesh_it != mesh_indices_order.end(); ++mesh_it)
+        {
+            const unsigned int mesh_idx = *mesh_it;
+            const Mesh& mesh = storage.meshgroup->meshes[mesh_idx];
+            const Point3 middle3 = mesh.getAABB().getMiddle();
+            const Point middle(middle3.x, middle3.y);
+            const coord_t dist2 = vSize2(middle - layer_start_position);
+            if (dist2 < best_dist2)
+            {
+                best_dist2 = dist2;
+                starting_mesh_it = mesh_it;
+            }
+        }
+    }
```

Comment:
Might be. Having separate functions is not a goal in itself.
I don't think this function is getting too long to understand without pulling this block into a new function.

Or do you think this function is less readable now?